test/wasm_test.cc

```diff
@@ -242,4 +244,84 @@ TEST_P(TestVm, AlwaysApplyCanary) {
   }
 }
 
+// Check that there are no stale thread-local cache keys (eventually)
+TEST_P(TestVm, CleanupThreadLocalCacheKeys) {
+  const auto *const plugin_name = "plugin_name";
+  const auto *const root_id = "root_id";
+  const auto *const vm_id = "vm_id";
+  const auto *const vm_config = "vm_config";
+  const auto *const plugin_config = "plugin_config";
+  const auto fail_open = false;
+
+  WasmHandleFactory wasm_handle_factory =
+      [this, vm_id, vm_config](std::string_view vm_key) -> std::shared_ptr<WasmHandleBase> {
+    auto base_wasm = std::make_shared<WasmBase>(makeVm(engine_), vm_id, vm_config, vm_key,
+                                                std::unordered_map<std::string, std::string>{},
+                                                AllowedCapabilitiesMap{});
+    return std::make_shared<WasmHandleBase>(base_wasm);
+  };
+
+  WasmHandleCloneFactory wasm_handle_clone_factory =
+      [this](const std::shared_ptr<WasmHandleBase> &base_wasm_handle)
+      -> std::shared_ptr<WasmHandleBase> {
+    auto wasm = std::make_shared<WasmBase>(
+        base_wasm_handle, [this]() -> std::unique_ptr<WasmVm> { return makeVm(engine_); });
+    return std::make_shared<WasmHandleBase>(wasm);
+  };
+
+  PluginHandleFactory plugin_handle_factory =
+      [](const std::shared_ptr<WasmHandleBase> &base_wasm,
+         const std::shared_ptr<PluginBase> &plugin) -> std::shared_ptr<PluginHandleBase> {
+    return std::make_shared<PluginHandleBase>(base_wasm, plugin);
+  };
+
+  // Read the minimal loadable binary.
+  auto source = readTestWasmFile("abi_export.wasm");
+
+  // Simulate a plugin lifetime.
+  const auto plugin1 = std::make_shared<PluginBase>(plugin_name, root_id, vm_id, engine_,
+                                                    plugin_config, fail_open, "plugin_1");
+  auto base_wasm_handle1 =
+      createWasm("vm_1", source, plugin1, wasm_handle_factory, wasm_handle_clone_factory, false);
+  ASSERT_TRUE(base_wasm_handle1 && base_wasm_handle1->wasm());
+
+  auto local_plugin1 = getOrCreateThreadLocalPlugin(
+      base_wasm_handle1, plugin1, wasm_handle_clone_factory, plugin_handle_factory);
+  ASSERT_TRUE(local_plugin1 && local_plugin1->plugin());
+  local_plugin1.reset();
+
+  auto stale_plugins_keys = internal::testing::staleLocalPluginsKeys();
+  ASSERT_EQ(1, stale_plugins_keys.size());
```

Comment:
nit: suggest using `EXPECT_EQ` here, since the test can continue on: http://google.github.io/googletest/primer.html#assertions